### PR TITLE
Add rakudo.eclass and its dependencies.

### PR DESCRIPTION
**Pull request by amano-kenji**
rakudo.eclass requires install-dist.raku from the new rakudo ebuild.

dev-raku/TAP, dev-raku/Getopt-Long, and dev-raku/App-Prove6 are necessary for running tests through rakudo.eclass.

### CODE REVIEW

**Comment by amano-kenji** [comment]
I just added a sign off.

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (thesamesam) [inline dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:14]: Why restrict?

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (thesamesam) [inline dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:21]: We prefer arrays for DOCS, so:
DOCS=( ... )

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (thesamesam) [inline dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:11]: New line before LICENSE.

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (thesamesam) [inline dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:8]:
```suggestion
DESCRIPTION="Run tests through a TAP harness"
```

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (thesamesam) [inline dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:10]: Could we/should we use the same SRC_URI magic as perl-module.eclass?

File: dev-raku/App-Prove6/metadata.xml
Comment (thesamesam) [inline dev-raku/App-Prove6/metadata.xml:3]: 1. https, not http
2. Missing any metadata at all, including maintainer?

File: eclass/rakudo.eclass
Comment (thesamesam) [inline eclass/rakudo.eclass:20]: We use Bash tests were possible (so [[ ... ]], not [ .. ] ), see https://projects.gentoo.org/qa/policy-guide/ebuild-format.html#pg0101.

File: eclass/rakudo.eclass
Comment (thesamesam) [inline eclass/rakudo.eclass:52]: Please define phases in order of execution (i.e. src_test should go before src_install).

File: eclass/rakudo.eclass
Comment (thesamesam) [inline eclass/rakudo.eclass:47]: Repeat for others:
```suggestion
	"${BROOT}"/usr/share/perl6/core/tools/install-dist.raku \
```

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (amano-kenji) [reply dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:14]: primaryuri is used in my personal overlay to not make unneccesary connections to gentoo servers.

Without primaryuri, src_fetch contacts gentoo servers to check for availability of source files.

I will try to address the rest as much as possible tomorrow.

Comment (thesamesam) [reply dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:14]: primaryuri doesn't make sense for something you're proposing for ::gentoo, though. So drop it for the PR?

Comment (amano-kenji) [reply dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:14]: I will soon.

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (amano-kenji) [reply dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:11]: Done.

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (amano-kenji) [reply dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:21]: Done.

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (amano-kenji) [reply dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:8]: Done

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (amano-kenji) [reply dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:10]: There are currently 3 different package ecosystems for raku.

I just read perl-module.eclass which already has 6 variables for handling cpan.

Beyond CPAN, raku also accepts packages from arbitrary git repositories and zef package ecosystem.

There is no algorithm for automatically generating SRC_URI for arbitrary git repositories.

I don't know how rakudo.eclass can automatically generate SRC_URI for zef package ecosystem. Zef download URLs include random numbers for each version of each package.

File: dev-raku/App-Prove6/metadata.xml
Comment (amano-kenji) [reply dev-raku/App-Prove6/metadata.xml:3]: 1. I replaced http with https.
2. In my overlay, I don't bother with package medata unless I have to describe USE flags. I had no problem with this approach in my personal overlay. What do you suggest for this?

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:20]: Done. All of it comes from maintaining my personal overlay.

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:52]: Done.

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:47]: Done for rakudo_src_compile and rakudo_src_install

**Comment by amano-kenji** [comment]
Anyone?

**Comment by mgorny** [comment]
We're waiting for you to resolve the CI failures. Also, please split commits.

**Comment by amano-kenji** [comment]
Because CI failures are common, I ignored it. I will make sure to correct CI failures and split commits tomorrow.

How should I split commits?

**Comment by amano-kenji** [comment]
I just split commits and fixed CI failures. Is it good now?

**Comment by amano-kenji** [comment]
Is anything blocking this pull request? Or, is it the case that everyone is busy?

File: dev-raku/App-Prove6/metadata.xml
Comment (thesamesam) [reply dev-raku/App-Prove6/metadata.xml:3]: I don't know. Who is going to look after this package? Are you?

File: dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild
Comment (thesamesam) [reply dev-raku/App-Prove6/App-Prove6-0.0.12-r2.ebuild:10]: OK, fair enough.

File: dev-lang/rakudo/rakudo-2022.12.ebuild
Comment (thesamesam) [inline dev-lang/rakudo/rakudo-2022.12.ebuild:1]: 1. You should use `pkgdev commit` - the copyright year is wrong here
2. The commit message format is also wrong (it should be `cat/pkg: add ...`)

Comment (amano-kenji) [reply dev-lang/rakudo/rakudo-2022.12.ebuild]: Is it good now?

File: dev-raku/App-Prove6/metadata.xml
Comment (amano-kenji) [reply dev-raku/App-Prove6/metadata.xml:3]: I think everyone including me. Everyone should take care of gentoo.

Comment (amano-kenji) [reply dev-raku/App-Prove6/metadata.xml:3]: What should be there if everyone should take the responsibility?

Comment (amano-kenji) [reply dev-raku/App-Prove6/metadata.xml:3]: I will take care of issues around rakudo.eclass as long as I use raku on gentoo. But, other people should also join.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (mgorny) [inline dev-lang/moarvm/moarvm-2022.12.ebuild:4]: Why EAPI 7?

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (mgorny) [inline dev-lang/moarvm/moarvm-2022.12.ebuild:22]: Please use standard variable order (as in `skel.ebuild`), at least for these variables.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (mgorny) [inline dev-lang/moarvm/moarvm-2022.12.ebuild:18]: Include the GitHub repo in HOMEPAGE (and keep it one per line), to make it easy to find sources.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (mgorny) [inline dev-lang/moarvm/moarvm-2022.12.ebuild:26]:
```suggestion
RDEPEND="
	dev-libs/libatomic_ops
	>=dev-libs/libuv-1.26:=
	dev-libs/libffi:=
"
```

Same style below.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (mgorny) [inline dev-lang/moarvm/moarvm-2022.12.ebuild:34]: RESTRICT goes immediately below IUSE.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (mgorny) [inline dev-lang/moarvm/moarvm-2022.12.ebuild:37]: Don't play with `MAKEOPTS`, pass them explicitly via `emake` instead.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (mgorny) [inline dev-lang/moarvm/moarvm-2022.12.ebuild:42]: Why would you einfo it? oO

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (mgorny) [inline dev-lang/moarvm/moarvm-2022.12.ebuild:49]: Why are you using flags to do that? If you really can't respect CC/CXX, why not use something like `tc-is-clang`? Also, I'm guessing the current logic will fail on clang-only systems w/ `USE=-clang`.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (mgorny) [inline dev-lang/moarvm/moarvm-2022.12.ebuild:33]:
```suggestion
# Tests are conducted in dev-lang/nqp ebuild
```

…if I'm guessing correctly what this was supposed to mean?

File: dev-lang/nqp/nqp-2022.12.ebuild
Comment (mgorny) [inline dev-lang/nqp/nqp-2022.12.ebuild:55]: But… how? `default` calls `default_src_prepare`, not anything java-specific.

File: dev-lang/nqp/nqp-2022.12.ebuild
Comment (mgorny) [inline dev-lang/nqp/nqp-2022.12.ebuild:62]:
```suggestion
	local myconfargs=(
		"--backend=${MULTIBUILD_VARIANT}"
		"--prefix=${EPREFIX}/usr"
	)
```

File: eclass/rakudo.eclass
Comment (mgorny) [inline eclass/rakudo.eclass:20]: This would really be much more readable as `case ${CATEGORY}/${PN} in` and setting some meaningfully named variable. Also, declare `BDEPEND` first, then perhaps just conditionally append to it? Also, `:=` is meaningless in `BDEPEND`, it's allowed only so you could do `BDEPEND=${RDEPEND}`.

Also, do these test deps really have no tests, or are you skipping them to avoid circular deps? If the latter, that's wrong.

File: eclass/rakudo.eclass
Comment (mgorny) [inline eclass/rakudo.eclass:31]: This is confusing. It makes it unclear whether you're supposed to pass just a filename or the whole path.

File: eclass/rakudo.eclass
Comment (mgorny) [inline eclass/rakudo.eclass:35]: Use relative paths.

File: eclass/rakudo.eclass
Comment (mgorny) [inline eclass/rakudo.eclass:39]: Any reason you need to call it via `env` rather than `local -x RAKUDO_RERESOLVE_DEPENDENCIES=0`?

**Review by mgorny — CHANGES_REQUESTED** [review]
You need to split new packages into their own commits. Also, I suggest taking rakudo over because with its current maintainer, it's not going to stay in Gentoo long.

File: dev-lang/rakudo/rakudo-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/rakudo/rakudo-2022.12.ebuild]: Resolving it for now.

File: dev-raku/App-Prove6/metadata.xml
Comment (amano-kenji) [reply dev-raku/App-Prove6/metadata.xml:3]: I will just add myself as the maintainer for all affected packages.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:4]: I didn't think much about it. I just inherited it from older ebuilds. I set it to 8 now.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:22]: Done by referring to https://devmanual.gentoo.org/ebuild-writing/variables/index.html

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:18]: Done.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:26]: I will address this and the rest tomorrow.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:49]: I am not using anything. I merely inherited everything from the previous maintainer, Patrick.

There is a saying, if it's not broken, don't fix it.

I will see what I can do about this.

File: dev-lang/nqp/nqp-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/nqp/nqp-2022.12.ebuild:55]: I didn't write it. Patrick did. I merely copied his ebuild. I will figure out what to do about this.

Comment (amano-kenji) [reply dev-lang/nqp/nqp-2022.12.ebuild:55]: My contribution to rakudo ecosystem is to add install-raku.dist and rakudo.eclass and rakudo.eclass test dependencies. I tried not to touch rakudo, nqp, and moarvm.

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:20]: The test deps run their own tests without prove6.

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:39]: I didn't know about `local -x`. I don't know much about bash. I only knew `env`.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:42]: I didn't do it. Patrick did.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:33]: I don't know. Patrick wrote it. I will look into it.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:26]: Done.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:34]: Done.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:37]: Done.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:42]: Removed.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:49]: I'm trying to address this, but I don't know what exactly you mean by CC, CXX, and `tc-is-clang`.

Do you mean that arguments to --libdir and --prefix should be handled by an existing eclass?

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:33]: Done as you asked.

File: dev-lang/nqp/nqp-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/nqp/nqp-2022.12.ebuild:55]: Should I just replace `default` with `java-pkg-opt-2_src_prepare`?

File: dev-lang/nqp/nqp-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/nqp/nqp-2022.12.ebuild:62]: Done.

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:20]: Can you show me how you would do it? I don't know much about bash.

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:31]: Changed
```
<executable-in-/usr/share/perl6/vendor/bin>
```
to
```
<executable-file-name-in-/usr/share/perl6/vendor/bin>
```

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:35]:
```
dosym "../share/perl6/vendor/bin/$1" "/usr/bin/$1" || die
```

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:39]: Done.

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass:20]: I just did it in my own way. See whether you like it.

**Comment by amano-kenji** [comment]
Note that I don't use JVM on my system. JVM backend for nqp and rakudo is not supported by me.

In my opinion, JVM programs should be distributed as uberjars. I really don't want to think about JVM packaging on gentoo linux.

Rakudo has the best story for integration with any POSIX-like operating system.

**Comment by amano-kenji** [comment]
I don't know what to do about
> metadata.xml: proxied maintainer missing proxy dev/project: Amano Kenji <amano.kenji@proton.me>

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:43]: Is `tc-is-clang` the right way?

Comment (mgorny) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:43]: There is also a saying "if it's Patrick, assume it's completely broken and wrong".

"The right way" would be to respect CC independently of value. However, I don't know what the option does. Can you tell me that?

File: dev-lang/nqp/nqp-2022.12.ebuild
Comment (mgorny) [reply dev-lang/nqp/nqp-2022.12.ebuild:55]: I suppose that was the intent.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:43]: Are you curious about --compiler option?

Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:43]: `perl Configure.pl --help` says
```
    --compiler <compiler>
        Currently supported compilers are "gcc", "clang" and "cl".
```

File: dev-lang/nqp/nqp-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/nqp/nqp-2022.12.ebuild:55]: Replaced `default` with `java-pkg-opt-2_src_prepare`.

File: dev-lang/moarvm/moarvm-2022.12.ebuild
Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:43]: `tc-get-compiler-type` may be able to replace clang USE flag although https://wiki.gentoo.org/wiki/Clang says

> Some packages are aware of the [USE](https://wiki.gentoo.org/wiki/USE_flag)=[clang](https://packages.gentoo.org/useflags/clang) USE flag. This is typically here as specific flags must be used by the package internally for it to compile or work correctly at runtime:

Comment (amano-kenji) [reply dev-lang/moarvm/moarvm-2022.12.ebuild:43]: I just replaced clang USE flag with `tc-get-compiler-type` in a case statement that sets `compiler` local variable.

File: eclass/rakudo.eclass
Comment (amano-kenji) [reply eclass/rakudo.eclass]: Resolving it for now.

**Comment by amano-kenji** [comment]
It seems rakudo-2023.02 already landed on gentoo linux. I need to update this pull request accordingly.

Anything more?

**Comment by mgorny** [comment]
I'm sorry but I can manage time to review this only every few days.

**Comment by amano-kenji** [comment]
Obviously, Patrick Lauer is back. So, I didn't change the maintainer for moarvm, nqp, and rakudo.

Please review again.

**Comment by mgorny** [comment]
The problem isn't whether he's around or not. The problem is that he's thoroughly incompetent and pretty proud of it.

**Comment by amano-kenji** [comment]
Finally, my pull request passed CI. Is it good now?

**Comment by amano-kenji** [comment]
I fixed build issues by building them inside my overlay. Is it good now?

**Comment by amano-kenji** [comment]
Is anyone here? I think it's already good enough to be merged. If there is any imperfection, it can be fixed later by someone else or me.

File: dev-lang/nqp/nqp-2023.02-r1.ebuild
Comment (vaukai) [inline dev-lang/nqp/nqp-2023.02-r1.ebuild:21]: dev-java/asm:4 is gone

Comment (amano-kenji) [reply dev-lang/nqp/nqp-2023.02-r1.ebuild:21]: There is dev-java/asm:9.

I'd rather exclude java from rakudo because

1. I don't use java on my system. Java backend is not going to be tested by me.
2. I think JVM is a dead end.
3. rakudo is not going to shine on JVM. It should run on MoarVM.

Comment (vaukai) [reply dev-lang/nqp/nqp-2023.02-r1.ebuild:21]: > Should we cut java out of rakudo?

Up to package maintainer, see https://gitweb.gentoo.org/repo/gentoo.git/commit/dev-lang/nqp?id=f94fd00ae0ca39fc553dd8e83f135c0004614e2b.

Comment (amano-kenji) [reply dev-lang/nqp/nqp-2023.02-r1.ebuild:21]: Okay, I will just let him deal with java backend. I will update accordingly.

Comment (amano-kenji) [reply dev-lang/nqp/nqp-2023.02-r1.ebuild:21]: Closing this to make a new pull request.

**Comment by amano-kenji** [comment]
Going to make a new pull request.

**Pull request closed without merging**
